### ISSUE TITLE
TypeScript definitions

### ISSUE THREAD

**Issue by tunnckoCore** [lukeed/sade#27]
Ooops, one more issue :D 

I may PR soon.

**Comment by tunnckoCore** [comment]
I'm not good at TS but something like that would work?

```ts
declare module 'sade' {
  interface CommandOptions {
    default: boolean;
  }
  interface ParseOptions {}
  type Action = (...args: any) => any;

  interface Sade {
    command(str: string, desc: string, opts: CommandOptions): Sade;
    describe(str: string): Sade;
    option(str: string, desc: string, val: any): Sade;

    action(handler: Action): Sade;

    example(str: string): Sade;

    version(str: string): Sade;

    parse(arr: string[], opts: ParseOptions): any;
  }
}
```

**Comment by lukeed** [comment]
For a project like sade, I would much rather they live inside DefintelyTyped, espeically since [`mri`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mri/index.d.ts)'s are already there.

The `sade` types can extend and require `mri`'s, which is the main reason for having them live there. If it were to live in-house, the types would be duplicated and probably inconsistent. Much better to have them resolve from the same place 👍 

If `sade` had no dependencies, I'd definitely ship them in this repo.

If you want to PR into DT, I'll help review and get them in!

**Comment by Epimodev** [comment]
Hello,
Is there any update about PR into DefintelyTyped ?
If no one plans to do it in a few days, I can make a PR.

**Comment by Epimodev** [comment]
Hello @lukeed 

I've just made a PR into DefintelyTyped for TypeScript users: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40398

I check the documentation and the source code to create it but I may have made some mistakes.
Can you review it, please?

**Comment by Epimodev** [comment]
The PR was already reviewed and merged by a DefinitelyTyped maintainer.
So we can download types from npm: https://www.npmjs.com/package/@types/sade.

It works well for me but I still interested in having a review from you.

**Comment by lukeed** [comment]
Closed via 9171cbf Sade will ship its own type definitions starting with `1.8.0`

Thank you to @Epimodev who filled in w/ the `@types/sade` package in the interim :) 

Most notable difference (I think) is that your commands' `action` handlers can now have their arguments typed directly, allowing you to save/reuse them. It also always injects the `mri.Argv` type for the last `options` argument:

```ts
sade('hello')
    .command('foobar <name> [age]', 'asd', {
        default: true
    })
    .action<[string, number|null]>((name, age, options) => {
        name.length;
        // ^? type: string
        age
        // ^? type: number | null
        options
        // ^? type: mri.Argv<Default>
    })
    // ...
```
